pkg/controller/sync/controller.go

```diff
@@ -373,3 +364,123 @@ func (s *FederationSyncController) syncToClusters(fedResource FederatedResource,
 
 	return status
 }
+
+func (s *FederationSyncController) ensureDeletion(fedResource FederatedResource) error {
+	fedResource.DeleteVersions()
+
+	key := fedResource.FederatedName().String()
+	kind := fedResource.FederatedKind()
+
+	glog.V(2).Infof("Ensuring deletion of %s %q", kind, key)
+
+	obj := fedResource.Object()
+
+	finalizers := sets.NewString(obj.GetFinalizers()...)
+	if !finalizers.Has(FinalizerSyncController) {
+		glog.V(2).Infof("%s %q does not have the %q finalizer. Nothing to do.", kind, key, FinalizerSyncController)
+		return nil
+	}
+
+	annotations := obj.GetAnnotations()
+	orphanResources := annotations != nil && annotations[OrphanManagedResources] == "true"
+	if orphanResources {
+		glog.V(2).Infof("Found %q annotation on %s %q. Removing the finalizer.", OrphanManagedResources, kind, key)
+		err := s.removeFinalizer(fedResource)
+		if err != nil {
+			return errors.Wrapf(err, "failed to remove finalizer %q", OrphanManagedResources)
+		}
+		glog.V(2).Infof("Initiating the removal of the label %q from resources previously managed by %s %q.", util.ManagedByFederationLabelKey, kind, key)
+		return nil
+		// TODO(marun) Implement removal of labels
+	}
+
+	glog.V(2).Infof("Deleting resources managed by %s %q from member clusters.", kind, key)
+
+	// First retrieve a list of unready clusters to maximize the
+	// changes of detecting when it's not possible to guarantee
+	// removal from member clusters.
+	unreadyClusters, err := s.informer.GetUnreadyClusters()
+	if err != nil {
+		return errors.Wrap(err, "failed to get a list of unready clusters")
+	}
+
+	clusters, err := s.informer.GetReadyClusters()
+	if err != nil {
+		return errors.Wrap(err, "failed to get a list of ready clusters")
+	}
+
+	updater := fedResource.NewUpdater()
+	deletedFromClusterNames := []string{}
+	targetKey := fedResource.TargetName().String()
+	retrievalErrorClusterNames := []string{}
+	for _, cluster := range clusters {
+		clusterName := cluster.Name
+		rawClusterObj, _, err := s.informer.GetTargetStore().GetByKey(clusterName, targetKey)
+		if err != nil {
+			runtime.HandleError(errors.Wrapf(err, "failed to retrieve managed resource for cluster %q", clusterName))
+			retrievalErrorClusterNames = append(retrievalErrorClusterNames, clusterName)
+			continue
+		}
+		// Already removed
+		if rawClusterObj == nil {
```

Comment:
i think this can also mean that the informer has not observed it yet.
isn't it safer to issue delete to confirm notfound?

Reply:
If the informer hasn't observed it yet, the resource will be reconciled again when it does observe it.  What would be safer about issuing a deletion on the basis of potentially incomplete information?

Reply:
Without the resource in the informer cache, it is entirely possible that the resource in the member cluster is not managed.  As per #549, only managed resources should be deleted.

Reply:
do you plan to make it delete labeled objects when federated object is not there? then it makes sense.

Reply:
Yes, that's the plan